WalletWasabi/Cache/MemoryCache.cs

```diff
@@ -0,0 +1,83 @@
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+
+namespace WalletWasabi.Cache
+{
+	public class MemoryCache<TKey, TValue> where TKey : notnull
+	{
+		private readonly record struct ExpirableValue
+		{
+			public readonly TValue Value;
+			public readonly long WhenToEvict;
+
+			public ExpirableValue(TValue value, TimeSpan expiration)
+			{
+				Value = value;
+				WhenToEvict = Environment.TickCount64 + (long)expiration.TotalMilliseconds;
+			}
+
+			public bool IsExpired => Environment.TickCount64 > WhenToEvict;
+		}
+
+		private readonly object _syncObj = new();
+		private readonly ConcurrentDictionary<TKey, ExpirableValue> _internalDictionary = new();
```

Comment:
All operations in `ConcurrentDictionary` are thread safe, however if you want to perform multiple operations, call multiple methods or a same method multiple times, you have to lock. `Clear` is thread safe and doesn't requires any lock because internally it locks everything.

Reply:
I know that but as I shared above, the idea is to continue building it.